src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java

```diff
@@ -1,127 +1,100 @@
 package com.teragrep.rlp_03.context.buffer;
 
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
 import java.nio.ByteBuffer;
-import java.util.concurrent.locks.Lock;
-import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.Phaser;
 
 public class BufferLeaseImpl implements BufferLease {
-    private static final Logger LOGGER = LoggerFactory.getLogger(BufferLeaseImpl.class);
-    private final long id;
-    private final ByteBuffer buffer;
-    private long refCount; // TODO consider using a semaphore
-    private final Lock lock;
-
-    public BufferLeaseImpl(long id, ByteBuffer buffer) {
-        this.id = id;
-        this.buffer = buffer;
-        this.refCount = 0;
-        this.lock = new ReentrantLock();
+    private final BufferContainer bufferContainer;
+    private final Phaser phaser;
+
+    private final BufferLeasePool bufferLeasePool;
+
+    public BufferLeaseImpl(BufferContainer bc, BufferLeasePool bufferLeasePool) {
+        this.bufferContainer = bc;
+        this.bufferLeasePool = bufferLeasePool;
+
+        this.phaser = new ClearingPhaser(1); // registered = 1
+    }
+
+    @Override
+    public BufferContainer bufferContainer() {
+        return this.bufferContainer;
     }
 
     @Override
     public long id() {
-        return id;
+        return this.bufferContainer.id();
     }
 
     @Override
     public long refs() {
-        lock.lock();
-        try {
-            return refCount;
-        }
-        finally {
-            lock.unlock();
-        }
+        // initial number of registered parties is 1
+        return phaser.getRegisteredParties();
     }
 
     @Override
     public ByteBuffer buffer() {
-        lock.lock();
-        try {
-            return buffer;
-        }
-        finally {
-            lock.unlock();
-        }
-
+        return this.bufferContainer.buffer();
     }
 
     @Override
     public void addRef() {
-        lock.lock();
-        try {
-            refCount++;
-        }
-        finally {
-            lock.unlock();
+        int a = this.phaser.register();
+        if (a < 0) {
```

Comment:
Might as well be inline as this variable is not used anywhere?
`if (this.phaser.register() < 0)`

Reply:
fix [c447182](https://github.com/teragrep/rlp_03/pull/84/commits/c44718257663e780cb7e9699b9c1159fcf9517e9)